plugins/qtosgrave/qtosgviewer.cpp

```diff
@@ -2112,21 +2127,34 @@ void QtOSGViewer::_UpdateEnvironment()
     }
 }
 
-void QtOSGViewer::_PostToGUIThread(const boost::function<void()>& fn, bool block)
+size_t QtOSGViewer::_GetQueueSizeForPriority(GUIThreadQueuePriority priority) {
+    size_t queueSize = 0;
+    for (uint8_t i = static_cast<uint8_t>(priority); i < 4; i++) {
+        queueSize += _mapGUIFunctionLists[static_cast<GUIThreadQueuePriority>(i)].size();
+    }
+    return queueSize;
+}
+
+void QtOSGViewer::_PostToGUIThread(const boost::function<void()>& fn, GUIThreadQueuePriority priority, bool block)
 {
     if( _nQuitMainLoop != -1 ) {
         // viewer quit, so anything posted won't get processed
         return;
     }
 
     boost::mutex::scoped_lock lockmsg(_mutexGUIFunctions);
-    if( _listGUIFunctions.size() > 1000 ) {
-        // can happen if system is especially slow
-        //RAVELOG_WARN_FORMAT("too many gui post commands, ignoring %d", _listGUIFunctions.size());
+    GUIThreadFunctionPtr pfn(new GUIThreadFunction(fn, block));
+    // Block non-essential fucntions if viewer is not processing any messages
+    if (!_bUpdateEnvironment && priority < GUIThreadQueuePriority::HIGH) {
+        RAVELOG_WARN("GUI thread function of priority %d dropped due to _bUpdateEnvironment being set to false", priority);
         return;
     }
-    GUIThreadFunctionPtr pfn(new GUIThreadFunction(fn, block));
-    _listGUIFunctions.push_back(pfn);
+    // GUI thread function queue size limit
+    if (_GetQueueSizeForPriority(priority) >= _mapGUIFunctionListLimits[priority]) {
+        RAVELOG_WARN("GUI thread function of priority %d dropped due to function queue for given priority exceeding size of %d", priority, _mapGUIFunctionListLimits[priority]);
+        return;
```

Comment:
Maybe dropping old command makes more sense?

For example, if someone is setting camera position, we would want to apply the new value